4.x - Add type hints to database package

markstory opened the pull request:
I had to move the part lists into Query::traverse() to maintain interface consistentcy with ExpressionInterface. Moving part lists also required a more clunky implementation for SQLServer, but that driver already has a few clunky parts.

A few other types needed to become nullable as we use them in a nullable way in test cases. I was concerned that removing those tests would break real-world usage.

Refs #11935

markstory committed:
Add strict_types to database package.

I had to widen the types on savepoint related methods as postgres uses
integer values for its save points.

markstory committed:
Add typehinting to database package.

I had to move the part lists into Query::traverse() to maintain
interface consistentcy with ExpressionInterface. Moving part lists
also required a more clunky implementation for SQLServer, but that
driver already has a few clunky parts.

A few other types needed to become nullable as we use them in a nullable
way in test cases. I was concerned that removing those tests would break
real-world usage.

ADmad commented on src/Database/Log/QueryLogger.php:
```diff
@@ -45,7 +45,7 @@ public function log(LoggedQuery $query)
      * @return void
      */
-    protected function _log($query)
+    protected function _log(\Cake\Database\Log\LoggedQuery $query): void
```
Need to add "use" statement instead. Same for `_interpolate()` below.

ADmad commented on src/Database/SqlserverCompiler.php:
```diff
@@ -62,7 +54,7 @@ class SqlserverCompiler extends QueryCompiler
      * @return string
      */
-    protected function _buildInsertPart($parts, $query, $generator)
+    protected function _buildInsertPart(array $parts, \Cake\Database\Query $query, \Cake\Database\ValueBinder $generator): string
```
FQCN need to be changed.

markstory committed:
Fix feedback from review.

Also add bool cast as DSN strings will have `1` or `'1'` not a bool.

markstory committed:
Make savepoint methods use consistent types.

markstory committed:
Fix more inconsistent types.

markstory committed:
fix more typing issues.

dereuromark commented on src/Database/Dialect/SqlserverDialectTrait.php:
```diff
@@ -335,32 +335,32 @@ public function schemaDialect(): BaseSchema
      * Returns a SQL snippet for creating a new transaction savepoint
      *
-     * @param string $name save point name
+     * @param string|int $name save point name
```
Why do we need to allow int here?

markstory replied:
Tests were failing as some of the db implementationa were paasing integers.

dereuromark replied:
I see, so casting beforehand is not feasible here.

dereuromark commented on src/Datasource/QueryInterface.php:
````diff
@@ -159,7 +159,7 @@ public function count(): int;
      * ```
      *
-     * @param int $num number of records to be returned
+     * @param int|mixed $num number of records to be returned
````
The mixed here sounds bad - as an addition it makes it pretty much worthless to expect and input type.

markstory replied:
Datasource doesn't contain a way to describe a QueryExpression which is what the database implementation can accept in addition to an int. Ither implementations may accept othe 'expression' style objects as well.

lorenzo commented on src/Database/Query.php:
```diff
@@ -277,12 +309,12 @@ public function sql(?ValueBinder $generator = null): string
     {
-        $parts = $parts ?: array_keys($this->_parts);
+        $partList = "_{$this->type()}Parts";
+        $parts = $this->{$partList} ?? array_keys($this->_parts);
```
I would really like to keep the parts in the query compiler, unless you think this is a better idea.

One way to keep that design is by adding a traverseParts() function to the query, with the signature and implementation this function had before.

markstory replied:
I am not convinced what I have is great. Adding a `traverseParts` method to Query sounds good to me.

markstory committed:
Update based on feedback.

Moving part lists to the Query didn't sit well with Jose. He suggested
adding a new method to traverse a subset of parts, so I did that.

markstory committed:
Fix inconsistent type causing php7.1 errors.
